servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectLBHttpConnectionFactory.java

```diff
@@ -0,0 +1,174 @@
+            final ExecutionStrategy connectStrategy,
+            final ConnectionFactoryFilter<ResolvedAddress, FilterableStreamingHttpConnection> connectionFactoryFilter,
+            final ProtocolBinding protocolBinding) {
+        super(config, executionContext, version -> reqRespFactory, connectStrategy, connectionFactoryFilter,
+                connectionFilterFunction, protocolBinding);
+        requireNonNull(config.h1Config(), "H1ProtocolConfig is required");
+        assert config.connectAddress() != null;
+        connectAddress = config.connectAddress().toString();
+    }
+
+    @Override
+    Single<FilterableStreamingHttpConnection> newFilterableConnection(final ResolvedAddress resolvedAddress,
+                                                                      final TransportObserver observer) {
+        assert config.h1Config() != null;
+        return buildStreaming(executionContext, resolvedAddress, config, observer)
+                .map(c -> new PipelinedStreamingHttpConnection(c, config.h1Config(),
+                        reqRespFactoryFunc.apply(HTTP_1_1), config.allowDropTrailersReadFromTransport()))
+                .flatMap(this::processConnect);
+    }
+
+    // Visible for testing
+    Single<FilterableStreamingHttpConnection> processConnect(final NettyFilterableStreamingHttpConnection c) {
+        try {
+            // Send CONNECT request: https://datatracker.ietf.org/doc/html/rfc9110#section-9.3.6
+            // Host header value must be equal to CONNECT request target, see
+            // https://github.com/haproxy/haproxy/issues/1159
+            // https://datatracker.ietf.org/doc/html/rfc7230#section-5.4:
+            //   If the target URI includes an authority component, then a client MUST send a field-value
+            //   for Host that is identical to that authority component
+            final StreamingHttpRequest request = c.connect(connectAddress).setHeader(HOST, connectAddress);
+            // No need to offload because there is no user code involved
+            request.context().put(HTTP_EXECUTION_STRATEGY_KEY, offloadNone());
+            return c.request(request)
+                    .flatMap(response -> {
+                        // Successful response to CONNECT never has a message body, and we are not interested in payload
+                        // body for any non-200 status code. Drain it asap to free connection and RS resources before
+                        // starting TLS handshake or propagating an error.
+                        if (response.status().statusClass() != SUCCESSFUL_2XX) {
+                            return drainPropagateError(response, new ProxyResponseException(c +
+                                    " Non-successful response from proxy CONNECT " + connectAddress, response.status()))
+                                    .shareContextOnSubscribe();
+                        }
+                        return response.messageBody().ignoreElements()
```

Comment:
looks like draining is done in both codepaths (error or success) - consider draining it beforehand and then check the status to potentially bail out?

Reply:
I considered that, but then realized that if draining logic throws, we will never verify the response status and won't generate `ProxyResponseException`. It feels that knowing the response status will give users more information for troubleshooting vs seeing some weird error as a result of payload body draining. Because if that, decided to always drain after. Will add a comment to clarify.

Reply:
👍